File: searchMatrix2.java
/*
Time complexity : O(m+n)
Space complexity: O(1)
*/
public class searchMatrix2 {
    public boolean searchMatrix(int[][] matrix, int target) {
        int row = 0;
        int col = matrix[0].length-1;
        /* traverse from top right corner and if the value is 
            greater than target, decrease the column else 
            increment the row.
        */
        while(col >= 0 && row < matrix.length) {
            if(matrix[row][col] > target) {
                col--;
            } else if(matrix[row][col] < target) {
                row++;
            } else {
                return true;
            }
        }
        return false;
    }
}

File: searchMatrix2.py
# Time Complexity: O(m+n)
# Space Complexity: O(1)

class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        m = len(matrix)
        n = len(matrix[0])
        row = 0
        col = n-1
        while (0<= row <m and 0<= col <n):
            if matrix[row][col] == target:
                return True
            elif matrix[row][col] > target:
                col -= 1
            else:
                row +=1
